mysys/lf_alloc-pin.c

```diff
@@ -256,8 +256,8 @@ static int ptr_cmp(void **a, void **b)
 #define add_to_purgatory(PINS, ADDR)                                    \
   do                                                                    \
   {                                                                     \
-    *(void **)((char *)(ADDR)+(PINS)->pinbox->free_ptr_offset)=         \
-      (PINS)->purgatory;                                                \
+    my_atomic_storeptr_explicit((void **)((char *)(ADDR)+(PINS)->pinbox->free_ptr_offset), \
+      (PINS)->purgatory, __ATOMIC_RELEASE);                             \
```

Comment:
as far as I remember, purgatory is strictly thread-local and never accessed by other threads. So it should not need any memory barriers, I presume.

Reply:
I thought the same way. But as you can see here it uses memory of `LF_SLIST::key`, which can be concurrently accessed at this point (deleted element, which is "being found" by another thread).

Reply:
An element cannot be found after it's deleted, because deletion removes it from all lists.
So the case is that one thread has found the element and at the same time a different thread has deleted it. Right? I think it is possible, indeed, but then `key` should not be used for linking `LS_SLIST` elements into a purgatory, because it violates the `free_ptr_offset` constraint, as you wrote in a different comment.

Reply:
@vuvova yes, I thought the same way. :)
But then I realised that everything is fine as soon as load order of `key` and `link` is preserved: if `key` is guaranteed to get loaded before `link`, we won't reach the code that makes use of `key`.

But then we have another problem discovered by @mbeck21. If we really end up with 
```
RLX R link
RLX R link
RLX R key
REL W pin
```

We're supposed to make sure that `link` didn't change while we've been busy pining it. Can it happen so that compiler eliminated this loop?

OMG, somebody removed `volatile` specifier from the `link` but missed to make these loads atomic.